python/ray/_private/resource_spec.py

```diff
@@ -275,17 +251,98 @@ def resolve(self, is_head: bool, node_ip_address: Optional[str] = None):
                 )
 
         spec = ResourceSpec(
-            num_cpus,
-            num_gpus,
-            memory,
-            object_store_memory,
-            resources,
-            redis_max_memory,
+            num_cpus, num_gpus, memory, object_store_memory, resources, redis_max_memory
         )
         assert spec.resolved()
         return spec
 
 
+def _detect_gpus(num_gpus: Optional[int], resources: dict) -> int:
+    """Detect GPUs by rules of 'Hardware Accelerators (GPUs) on Ray', link:
+    https://github.com/ray-project/ray/blob/master/python/ray/util/accelerators/accelerators.md
+    The mainly rule is Homogenous within a node:
```

Comment:
Better to rephrase like: The GPU type in the same node should be the same, but different node can have different types of GPUs.